stl/inc/xstring

```diff
@@ -2526,38 +2516,34 @@ private:
         _Apply_annotation(_My_data._Myptr(), _My_data._Myres, _My_data._Mysize, _My_data._Myres);
     }
 
-    _CONSTEXPR20 void _Modify_annotation(const difference_type _Count) const noexcept {
-        // Extends/shrinks the annotated range by _Count
-        if (_Count == 0) {
-            return;
-        }
-
+    _CONSTEXPR20 void _Modify_annotation(const size_type _Old_size, const size_type _New_size) const noexcept {
         auto& _My_data = _Mypair._Myval2;
-        _Apply_annotation(
-            _My_data._Myptr(), _My_data._Myres, _My_data._Mysize, static_cast<size_type>(_My_data._Mysize + _Count));
+        _Apply_annotation(_My_data._Myptr(), _My_data._Myres, _Old_size, _New_size);
     }
 
-    _NODISCARD static const void* _Get_aligned_first(const void* _First, const size_type _Capacity) noexcept {
-        const char* _CFirst = reinterpret_cast<const char*>(_First);
+    _CONSTEXPR20 void _Annotate_switch_to_large() const noexcept {
+        // annotates exactly the pointer as valid
+        _Remove_annotation();
 
-        if (_Capacity >= _Asan_granularity) { // We are guaranteed to have sufficient space to find an aligned address
-            return reinterpret_cast<const void*>(
-                (reinterpret_cast<uintptr_t>(_CFirst) + (_Asan_granularity - 1)) & ~(_Asan_granularity - 1));
-        }
+        auto& _My_data              = _Mypair._Myval2;
+        const void* const _My_buf   = _My_data._Bx._Buf;
+        const void* const _Ptr_last = &_My_data._Bx._Ptr + 1;
+        const void* const _Buf_last = _My_data._Bx._Buf + _BUF_SIZE;
 
-        uintptr_t _Alignment_offset = reinterpret_cast<uintptr_t>(_CFirst) & (_Asan_granularity - 1);
-        if (_Alignment_offset != 0) {
-            _Alignment_offset = _Asan_granularity - _Alignment_offset;
-        }
+        constexpr bool _Small_string_always_aligned = _Memcpy_val_offset % _Asan_granularity == 0;

```

Comment:
Doesn't this assume that `*this` is always aligned to `_Asan_granularity` (8), which I think is true only on 64-bit platforms? (Also on 2566. If this were a `private: static constexpr` class member we'd only need to fix it once =))

Reply:
This checks if `*this` is always aligned to `_Asan_granularity`; yeah, it'll be true only on 64-bit platforms.

Reply:
Per our offline discussion: this validates that the offset of the SSO buffer within `*this` is a multiple of `_Asan_granularity`, but it might still be misaligned if `this` itself is not aligned to a multiple of `_Asan_granularity`.